### ISSUE TITLE
Ability to Configure Custom Logback Encoder/Layout

### ISSUE THREAD

**Issue by ajmath** [dropwizard/dropwizard#1388]
I would like the ability to log output in JSON to help with ingesting them into an ELK setup.  This can be accomplished by adding a custom Encoder or Layout on the appender.  The current way of configuring appenders doesn't expose this and so it must be done manually at run-time.

I don't have a strong opinion but I feel like either layout or encoder would be fine, and that we shouldn't use both.  I see that there is already a layout being passed into `AppenderFactory.build` which may create some issues.

Ideally, we could just pass in a class name with a no-arg constructor as a string parameters in the appenders config section.

Would a PR that added this feature be welcome?

**Comment by ajmath** [comment]
Any feedback on this?  I'm open to doing the work but I want to make sure that it is something that will be accepted.

**Comment by brentryan** [comment]
I did this by just implementing my own appender that extends from the current file appender provided by dropwizard. Then I pass in my custom json layout and formatters. Everything works great.

**Comment by ajmath** [comment]
@brentryan will that also work for the logger being used by Jetty that logs basic HTTP request events?

**Comment by ajmath** [comment]
Nevermind on how to configure it.  I saw an example in ConsoleAppenderFactory.java

**Comment by ajmath** [comment]
For the curious, here is what I ended up doing to make this work.

https://gist.github.com/ajmath/e9f90c29cd224653c218

My only gripe is having to add the file META-INF/services file that contains my class name.  Is there a way to have this done automatically via reflection?

**Comment by jplock** [comment]
Does https://github.com/dropwizard/dropwizard/pull/1415 address this?

**Comment by mwhipple** [comment]
@ajmath To answer the question about having to use the ServiceLoader META-INF/services file: the issue is that classes normally aren't loaded until they are needed in Java ([spec](http://docs.oracle.com/javase/specs/jls/se8/html/jls-12.html#jls-12.4.1)) so there needs to be some reference pointing to the class to have it picked up. There are different ways to do it but the ServiceLoader is standard and there's no good automatic way due to the way classes are loaded.

**Comment by jamisonhyatt** [comment]
@jplock I think those changes made it more difficult to do what @ajmath did in the gist he linked.  I commented on this issue here, https://github.com/dropwizard/dropwizard/issues/1451

**Comment by ragebiswas** [comment]
@ajmath Thanks for the gist (https://gist.github.com/ajmath/e9f90c29cd224653c218) - very helpful. One issue I see is that when I generate a dropwizard uber-jar with the maven shade plugin, the manifest file gets overwritten and dropwizard is unable to find the custom appender. Did you face this problem?

**Comment by ajmath** [comment]
@ragebiswas Unfortunately we aren't using the shade plugin so I didn't hit this problem.  Though, I remember a similar problem from a while back when I tried using that plugin.  I believe you can configure have it exclude certain files or force a certain version of a file in the result jar.

**Comment by heldeen** [comment]
@ragebiswas, the maven shade plugin has support for rewriting the manifest with the relocated classes. See https://maven.apache.org/plugins/maven-shade-plugin/examples/resource-transformers.html#ServicesResourceTransformer

**Comment by mhurne** [comment]
I know there's always a debate to be had about this sort of change because Dropwizard aims to be opinionated. IMO, this is a case where the opinion(s) should be defaults, but customization should be easy to achieve via configuration. Otherwise, we all customize it anyway, but have to do so by jumping through hoops.

Maybe something like the following should be possible:

```
  appenders:
    - type: file
      layout:
        type: json
```

I acknowledge that I may not be aware of reasons to avoid supporting this, or reasons why it may be difficult to achieve...

**Comment by joschi** [comment]
Closed via #2232
